src/handler.rs

```diff
@@ -577,9 +609,239 @@ where
+
+        self.set_inflight_region_desc_num(inflight.num_queues, inflight.queue_size);
+
+        Ok(())
+    }
 }
 
-impl<S, V, B: Bitmap> Drop for VhostUserHandler<S, V, B> {
+impl<S, V, B> VhostUserHandler<S, V, B>
+where
+    S: VhostUserBackend<V, B>,
+    V: VringT<GM<B>>,
+    B: NewBitmap + Clone,
+{
+    fn get_inflight_queue_size(&mut self, queue_size: u16) -> usize {
+        let queue_region_size;
+        let descr_state_size;
+        let virtio_features = self.get_features().unwrap();
+
+        if virtio_features & (1 << VIRTIO_F_RING_PACKED) == 0 {
+            // Use descriptor and queue states for split virtqueues
+            queue_region_size = mem::size_of::<QueueRegionSplit>();
+            descr_state_size = mem::size_of::<DescStateSplit>();
+        } else {
+            // Use descriptor and queue states for packed virtqueues
+            queue_region_size = mem::size_of::<QueueRegionPacked>();
+            descr_state_size = mem::size_of::<DescStatePacked>();
+        }
+        queue_region_size + descr_state_size * queue_size as usize
+    }
+
+    fn memfd_alloc(&self, mmap_size: usize) -> VhostUserResult<(*mut c_void, i64, i64)> {
+        let mut ret_val;
+        let inflight_file;
+        let dup_inflight_file;
+
+        // create an anonymous file and return a file descriptor to it.
+        // unsafe as extern syscall funciton is called
+        ret_val = unsafe {
+            libc::syscall(
+                libc::SYS_memfd_create,
+                &std::ffi::CString::new("inflight-region").unwrap(),
+                libc::MFD_ALLOW_SEALING,
+            )
+        };
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdCreateError);
+        }
+
+        inflight_file = ret_val;
+
+        // truncate the inflight_file to "mmap_size" length.
+        // unsafe as extern "ftruncate" is called
+        ret_val = unsafe { libc::ftruncate(inflight_file as RawFd, mmap_size as i64) } as i64;
+
+        if ret_val == -1 {
+            return Err(VhostUserError::FileTrucateError);
+        }
+
+        // place seals to restrict further modifications to the inflight_file file.
+        // unsafe as extern fcntl is called.
+        ret_val = unsafe {
+            libc::fcntl(
+                inflight_file as RawFd,
+                libc::F_ADD_SEALS,
+                libc::F_SEAL_GROW | libc::F_SEAL_SHRINK | libc::F_SEAL_SEAL,
+            )
+        } as i64;
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdSealError);
+        }
+
+        // duplicate the inflight_file file descriptor.
+        // unsafe as extern dup is called.
+        ret_val = unsafe { libc::dup(inflight_file as RawFd).into() };
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdCreateError);
+        }
+
+        dup_inflight_file = ret_val;
+
+        // map the inflight_file into memory.
+        // unsafe as extern mmap is called
+        Ok((
+            unsafe {
```

Comment:
We need to handle the case that mmap fails.

Reply:
I see, makes sense. Perhaps we should add a new error enum into vhost_user::Error?

Reply:
It could be solved/refined by refactor the code. The suggestion is to avoid RawFd if possible, because it doesn't solve the file descriptor ownership. The way to avoid RawFd is to convert it into File as soon as possible. And it's safe and easy to get RawFd from File object by file.as_raw_fd().